### PR TITLE
Add LTREB Prairie-forest ecotone of eastern Kansas/Foster Lab dataset

### PR DESCRIPTION
**Pull request by coolalexzb**
Hi, I have added  prairie-forest datasets from LTREB metioned in #1138 . If any mistake, please let me know and I will correct. Thank you!

### CODE REVIEW

**Comment by coolalexzb** [comment]
I added LTREB Prairie-forest dataset to data retriever successfully. Now, could we discuss further details after this step?  Thanks! @henrykironde

File: scripts/prairie_forest.json [hunk 1]
```diff
@@ -0,0 +1,1437 @@
...
+                    }
+                ]
+            },
+            "url": "https://foster.ku.edu/sites/foster.ku.edu/files/files/E2%20Plant%20Species%20Composition%206%2016.csv"
+        }
+    ],
+    "retriever": "True",
+    "retriever_minimum_version": "2.0.dev",
+    "title": "Prairie-forest ecotone of eastern Kansas/Foster Lab",
+    "urls": {
```
Comment (zhangcandrew) [inline scripts/prairie_forest.json]: Delete this field

**Comment by zhangcandrew** [comment]
Thanks @coolalexzb!

The script seems to be failing for me. You can test it out by running retriever install csv prairie-forest to make sure it install correctly. 
image

You should also run python version.py from your local retriever repo folder, to update the version.txt file.

File: scripts/prairie_forest.json [hunk 2]
```diff
@@ -0,0 +1,1437 @@
...
+                        "name": "N",
+                        "type": "int"
+                    },
+                    {
+                        "name": "P",
+                        "type": "int"
+                    },
+                    {
+                        "name": "Live_Biomass",
+                        "type": "float"
```
Comment (zhangcandrew) [inline scripts/prairie_forest.json]: we don't support float, the script type we use should be double

File: scripts/prairie_forest.json (hunk 1 above)
Comment (zhangcandrew) [reply scripts/prairie_forest.json]: We no longer use the urls field in our scripts. (#1089)

File: scripts/prairie_forest.json (hunk 2 above)
Comment (coolalexzb) [reply scripts/prairie_forest.json]: I completed changes that you suggested. The download function works well while the install function still has the previous problem. I checked elements of table again but did not find any problem. I feel it difficult to find the bug and correct it. @zhangcandrew

**Comment by henrykironde** [comment]
@coolalexzb, thanks for to the effort in hunting down the bug. 
I am using `Pycharm` for `debugging`. 
After running the PR with debug mode in the terminal 
`retriever install --debug sqlite prairie-forest`, I got this error.
```python
retriever install --debug sqlite prairie-forest
=> Installing prairie-forest
'>' not supported between instances of 'str' and 'int'
Traceback (most recent call last):
  File "/Users/henry/GitHub/Environments/py36/bin/retriever", line 11, in <module>
    sys.exit(main())
  File "/Users/henry/GitHub/Environments/py36/lib/python3.6/site-packages/retriever/__main__.py", line 199, in main
    dataset.download(engine, debug=debug)
  File "/Users/henry/GitHub/Environments/py36/lib/python3.6/site-packages/retriever/lib/templates.py", line 119, in download
    self.process_tables(table_obj, url)
  File "/Users/henry/GitHub/Environments/py36/lib/python3.6/site-packages/retriever/lib/templates.py", line 155, in process_tables
    self.engine.auto_create_table(table_obj, url=url)
  File "/Users/henry/GitHub/Environments/py36/lib/python3.6/site-packages/retriever/lib/engine.py", line 221, in auto_create_table
    if self.table.header_rows > 0 and not self.table.columns:
TypeError: '>' not supported between instances of 'str' and 'int'
```
In Pycharm, I set up break points at the last function `auto_create_table`. And the values of `if self.table.header_rows > 0 and not self.table.columns` show that it is a string "1" for self.table.header_rows. This is being compared to > 0. I realized that in your json file, you have given 
```json
            "dialect": {
                "delimiter": ",",
                "header_rows": "1"
            },
```
It should be 

```json
            "dialect": {
                "delimiter": ",",
                "header_rows": 1
            },
```
Or even removing the header value is fine. the default is assumed to be one in conventional tabular data files

**Comment by coolalexzb** [comment]
Thank you for helping me debug and the new json is correct. Now can we discuss futher details about the idea "Retriever Out of Memory Functionality"? Thanks! @henrykironde

**Comment by henrykironde** [comment]
Sorry @coolalexzb, we got a bit busy and forgot this. However I have run `retriever install sqlite prairie-forest`. and failed because there are some wrong columns.

**Comment by coolalexzb** [comment]
I find the problem is that the source csv file contains two columns containing the same title "Unknown species C", and the value of the two column is not all the same. Could you provide regular solution to such kind of problem?  Thanks! @henrykironde 
DeepinScrot-5948
DeepinScrot-5840

File: scripts/prairie_forest.json [hunk 3]
```diff
@@ -0,0 +1,1431 @@
...
+                    },
+                    {
+                        "name": "Unknown grass A",
+                        "type": "double"
+                    },
+                    {
+                        "name": "Unknown species C",
+                        "type": "double"
+                    },
+                    {
```
Comment (henrykironde) [inline scripts/prairie_forest.json]: We can rename the two as
unknown_species_c_first
unknown_species_c_second

Also, remove spaces from all the columns by replacing the spaces with _

**Comment by coolalexzb** [comment]
I replace all the space in the colunms with _ and command `retriever install sqlite prairie-forest` works well. I also want to know the whole steps of testing adding datasets, so I can test by myself when I need to add datasets to retriever next time. 
Besides, I want to share you something about the idea "Retriever Out of Memory Functionality". I add code to acquire the size of datasets before downloading, but I am still test the threshold of large datasets. So I cannot pull a request now. I also test more datasets to find the threshold and I find that getting the sizes of all the datasets in retriever is necessary. Now, I am trying to acquire size of datasets in retriever before downloading, especially the size after unzipping. I recommend that adding size of data as part of property of datasets to distinguish large datasets and I will complete this work. Thanks! @henrykironde

File: scripts/prairie_forest.json (hunk 3 above)
Comment (coolalexzb) [reply scripts/prairie_forest.json]: I replace all the space in the colunms with _ and command `retriever install sqlite prairie-forest `works well. I also want to know the whole steps of testing adding datasets, so I can test by myself when I need to add datasets to retriever next time.
Besides, I want to share you something about the idea "Retriever Out of Memory Functionality". I add code to acquire the size of datasets before downloading, but I am still test the threshold of large datasets. So I cannot pull a request now. I also test more datasets to find the threshold and I find that getting the sizes of all the datasets in retriever is necessary. Now, I am trying to acquire size of datasets in retriever before downloading, especially the size after unzipping. I recommend that adding size of data as part of property of datasets to distinguish large datasets and I will complete this work. Thanks! @henrykironde

Comment (henrykironde) [reply scripts/prairie_forest.json]: > whole steps of testing adding datasets, so I can test by myself when I need to add datasets to retriever next time.

we do not have any tools that test the dataset script. We are currently working on one however.
To manually test I usually run retriever on install on all the engines and look at the data if it is consistent in all engines. 

>  I add code to acquire the size of datasets before downloading, but I am still test the threshold of large datasets.

sounds good, let me know when you are ready for me to look at it

**Comment by henrykironde** [comment]
@coolalexzb Thanks for the PR. Merging

**Pull request merged**